indra/newview/llfloateremojipicker.cpp

```diff
@@ -631,9 +674,186 @@ BOOL LLFloaterEmojiPicker::handleKeyHere(KEY key, MASK mask)
 // virtual
 void LLFloaterEmojiPicker::closeFloater(bool app_quitting)
 {
+    saveState();
     LLFloater::closeFloater(app_quitting);
     if (mFloaterCloseCallback)
     {
         mFloaterCloseCallback();
     }
 }
+
+void LLFloaterEmojiPicker::onEmojiUsed(llwchar emoji)
+{
+    // Update sRecentlyUsed
+    auto itr = std::find(sRecentlyUsed.begin(), sRecentlyUsed.end(), emoji);
+    if (itr == sRecentlyUsed.end())
+    {
+        sRecentlyUsed.push_front(emoji);
+    }
+    else if (itr != sRecentlyUsed.begin())
+    {
+        sRecentlyUsed.erase(itr);
+        sRecentlyUsed.push_front(emoji);
+    }
+
+    // Increment and reorder sFrequentlyUsed
+    auto itf = sFrequentlyUsed.begin();
+    while (itf != sFrequentlyUsed.end())
+    {
+        if (itf->first == emoji)
+        {
+            itf->second++;
+            while (itf != sFrequentlyUsed.begin())
+            {
+                auto prior = itf;
+                prior--;
+                if (prior->second > itf->second)
+                    break;
+                prior->swap(*itf);
+                itf = prior;
+            }
+            break;
+        }
+        itf++;
+    }
+    // Append new if not found
+    if (itf == sFrequentlyUsed.end())
+        sFrequentlyUsed.push_back(std::make_pair(emoji, 1));
+}
+
+void LLFloaterEmojiPicker::loadState()
+{
+    if (!sStateFileName.empty())
+        return; // Already loaded
+
+    sStateFileName = gDirUtilp->getExpandedFilename(LL_PATH_USER_SETTINGS, "emoji_floater_state.xml");
+
+    llifstream file;
+    file.open(sStateFileName.c_str());
+    if (!file.is_open())
+    {
+        LL_WARNS() << "Emoji floater state file is missing or inaccessible: " << sStateFileName << LL_ENDL;
+        return;
+    }
+
+    LLSD state;
+    LLSDSerialize::fromXML(state, file);
+    if (state.isUndefined())
+    {
+        LL_WARNS() << "Emoji floater state file is missing or ill-formed: " << sStateFileName << LL_ENDL;
+        return;
+    }
+
+    sSelectedGroupIndex = state[sKeySelectedGroupIndex].asInteger();
```

Comment:
Looks like if file loading failes, value will be random? This one likely needs to be initialized with some default value.

Reply:
The prefix "s" is used for file-scoped variables as well
![image](https://github.com/secondlife/viewer/assets/124201357/5b510e9a-af7e-4401-ba12-9128f6e91ac0)

Reply:
> Looks like if file loading failes, value will be random? This one likely needs to be initialized with some default value.

Usually static variables are being initialized by zero
Anyway, the note makes sense; fixed

Thanks